### ISSUE TITLE
Edit chat messages

### ISSUE THREAD

**Issue by sunjam** [nextcloud/spreed#1836]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

Related to #774, but a different feature request.  **I'd like the ability to edit my chat messages** in Talk.  This is a common practice when dealing with typos, etc. and an expectation of common users.

Thanks for reading, and thanks for all your great work!

**Comment by nickvergessen** [comment]
That should work only in a limited case, otherwise it get's out of control:
* No command was triggered
* No mention was sent
* No participant read the message already

**Comment by DudleyDursley** [comment]
Would it be possible to let the admin decide this? So that you can deactivate editing/deleting messages all together, maybe set a time limit or let the users do it whenever they want.

**Comment by tiotrom** [comment]
I second that! It would be very useful!

**Comment by pasouf** [comment]
Yes ! now about the limited cases mentionned by @nickvergessen , editing is a common feature in "enterprise grade" chats, because you have to use a proper language and that means editing, ultimately if you feel like implementing those limits is mandatory, then you should also allow us to untick those limits in the configuration.

**Comment by Nyco** [comment]
Most messengers that offer the "edit/correct message" feature:
* allow this whatever the admin decides
* allow this but only for the user's last sent message (not the ones sent nefore)
* some messangers allow receivers to see the sender's original and edited messages

**Comment by agentlibre** [comment]
I believe this is an underestimated issue, which has a bad impact on a wider adoption for NextCloud talk. For several members within my organization, the lack of an editing/deleting feature in a direct message app is simply a no-go; I can't convince them that we should adopt NextCloud Talk for this single reason (despite other advantages, such as the integration with NextCloud Deck). 😥

For the record, I've found this issue mentioned here as well: https://help.nextcloud.com/t/talk-edit-just-sent-message-share-local-files/72756 🤔

I hope something could be done about this rather promptly, this would be so great! 🤩

**Comment by nickvergessen** [comment]
> I hope something could be done about this rather promptly

Not the best way to communicate 🙈

I get the point, but for Person A having a grid view in the call is the most important thing, for Person B being able to moderate a call is the most important thing and for you it's editing of messages.

Please note that until a month ago the orientation line was more the chat messengers out there like Whatsapp, Threema, Signal, … and now the comparison line just expended quickly with Teams, Slack, Jitsi, … because of corona.
Feel free to help out and get involved. Any help counts, not only coding. But until then the items we can work on are rather limited.

**Comment by agentlibre** [comment]
> > I hope something could be done about this rather promptly
> 
> Not the best way to communicate see_no_evil

Yes, sorry, I've noticed this once the message was published and was happy to have... an editing option! 🤣

> I get the point, but for Person A having a grid view in the call is the most important thing, for Person B being able to moderate a call is the most important thing and for you it's editing of messages.

Sure, sure, just wanted to share this testimony because, in my case, this single factor has been decisive for some common (and rather Nextcloud-friendly) users.

> Please note that until a month ago the orientation line was more the chat messengers out there like Whatsapp, Threema, Signal, … and now the comparison line just expended quickly with Teams, Slack, Jitsi, … because of corona.
> 
> Feel free to help out and get involved. Any help counts, not only coding. But until then the items we can work on are rather limited.

I perfectly understand the situation. I have no tech skills but I'm actively advocating FOSS and, first and foremost NextCloud, which is the best "gateway drug" in my opinion (also thanks to the AGLPv3  choice in addition to the wonderful software it is). That's precisely why I'm raising this issue: I would like to be able to convince more people to come on board (and truth be told, it could scale up very fast from where I'm standing) but I need the arguably basic features to be in place to do so.

I don't want to be a party pooper though: I greatly appreciate the work that has been and is being done!

**Comment by jasondaigo** [comment]
I used rocketchat mattermost and matrix just now during corona times. I only use nextcloud on private installation. Not implementing edit or delete seem a bit stubborn to me. In a Sega / Blackberry way (huge fan though) 🙂

**Comment by nickvergessen** [comment]
> Feel free to help out and get involved. Any help counts, not only coding. But until then the items we can work on are rather limited.

**Comment by graue70** [comment]
> That should work only in a limited case, otherwise it get's out of control:
> 
>     * No command was triggered
>     * No mention was sent
>     * No participant read the message already

I know nothing about the technical difficulties involved but it would annoy me if I couldn't fix the bad typo in my last message only because I mentioned somebody in it.

> Feel free to help out and get involved.

If I wanted to implement this, would it be enough to change the string in the database or is there something else that needs to be done?

**Comment by nickvergessen** [comment]
See https://github.com/nextcloud/spreed/pull/4861#issuecomment-753967049

**Comment by graue70** [comment]
> 2. Since the last message id of a conversation is set and the messages are downloaded by our mobile clients, deleting from the DB is not a good idea and doesn't help a lot. Instead we should only wipe the content of the message (if the database structure allows it to be empty) and at the same time we need to add a "hidden" message that the message with ID X was deleted, so the clients can remove the message from their local store as well.

To clarify my question: Is this comment also true for editing messages? If so, wouldn't it be a better solution to have some kind of event telling clients to reload a specific message from the database?

Or in other words: Do you think your suggestion is the best solution or only the easiest to achieve?

**Comment by AxelRothe** [comment]
I second this motion.

We recently switched from Slack to Talk and for the most part are very happy. But the lack of editing a posts, makes it hard to create pinned or dynamic posts. THINK BOTS! 🤖

A bot needs to be able to change their post. Otherwise the Bot might spam small posts for every reply instead of simply updating a post that leads to a better overview. i.e. a daily stand up bot, that can post a list of users that have not submitted a stand up and update it as users add their posts.

We moved to Nextcloud Talk for the ability to actively help development. Is there a way for us to help with this issue? We have developers.

**Comment by nickvergessen** [comment]
There is an easier part (modifying the server + web UI) and a harder part (making sure all the mobile clients reflect this correctly as well). I guess looking at the PR of "Delete messages" can give quite a good view and template for edits. But then we need to make sure mobile clients also understand the new system message and update the old messages accordingly.

**Comment by dtcg** [comment]
Any updates on this?

**Comment by edmundlaugasson** [comment]
Is Talk message editing now possible with Nextcloud 22?

**Comment by Normand-Nadon** [comment]
I can't find the option in my instance of Nextcloud. Is this feature coming? Where does it stand in the priority list?
I can simply not convert my business to use Talk vs Teams because of that one missing bit... the rest is A-Ok !

Have a nice day

**Comment by nickvergessen** [comment]
It's not on our todo list at the moment, as we are currently focusing on calls and meetings

**Comment by edmundlaugasson** [comment]
Unfortunately still not implemented even in great [23th version](https://nextcloud.com/blog/nextcloud-hub-2-brings-major-overhaul-introducing-nextcloud-office-p2p-backup-and-more/) :(

**Comment by sunjam** [comment]
I'll follow up here having just read [this forum comment](https://help.nextcloud.com/t/nextcloud-hub-ii-brings-major-overhaul-introducing-nextcloud-office-p2p-backup-and-more/128244/12) and watching the 23 announcement video where chat editing is discussed.

#### What is the most compelling reason to support message editing? (from the video)

Every chat platform I use supports editing: **Matrix, Slack, Discord, Rocket.chat, IRC**, etc. over the last couple years.
I've personally bridged all of these platforms on a per room basis via Matrix and editing is supported across all participants on all platforms.

So, this is something I'm just used to having.  No worries on if it isn't implemented; I just won't be migrating away without it.

**Comment by sunjam** [comment]
Nope, no change

**Comment by Normand-Nadon** [comment]
I know it is an issue tracker and not a chat forum here, but it amazes me that this issue is not taken care of since 2009 and is the no-go for many real enterprise level customer... I tried to start a migration project for my company and it was one of the top 3 deciding factor not to do it! In fact, it was the nail in the coffin (the two other were change management for the employees and some issues related to our use of PowerBi that could both be overcomed)

Now, more constructive comment, how I envision the feature:
**Feature:**
Ability to edit a message, it is self explanatory

**Administration of said feature**
Maybe add a set of rules for messages editing:
- edit messages feature -> off / on
- Message can be edited after they were read -> no / yes
- Maximum delay before message editing -> no / yes (enter time delay in minutes)
- _Some other options related to message linking and interractions with other apps to avoid breaking the system (can't tell about those one with my current knowledge)_

Have a nice day all :)

**Comment by sunjam** [comment]
I do wonder about editing in regards to Matterbridge support in Talk, since
all other modern chat platforms do include edit functionality.  Edits are
common in those platforms, which does lead to frustration if bridged to irc.

**Comment by Normand-Nadon** [comment]
Could a developer on the project respond to the thread please just to let us know the status of this feature request.

- Is it considered or not?
- Is it planned?
- Is it parked to deal with later?

Just please inform the community about this feature, because as I said, and other did, it is the only show-stopper remaining for adoption in our business ! This single features makes it that we can't consider migrating the entire business to Nextcloud!

Good say to the team, and keep-up with the amazing work you do!

**Comment by edmundlaugasson** [comment]
Certainly sent message editing is crucial. But I would say, that also sending without sound and scheduling messages are also important, e.g. like in Telegram. Also videoconversation recording - if we have cloud storage as Nextcloud has, then why we cannot record conversations? Besides, I had issues with Talk in case of more participants. If there are tens of users, then already will be issues with sound, screen sharing doesn't work, etc. Also I experienced sharing is challenging, e.g. simple folder sharing but not only - also other resources sharing is challenging, other participants just cannot see these shares. In case of AD you need to know exact ID and cannot use real name. Also sync between AD does not work - user has to be logged at least once in and then can be shared. These are my experiences in larger organization with Nextcloud.

**Comment by yjjoe-head** [comment]
bump!
any approximation ETA for that issue?

**Comment by benfavre** [comment]
Need this

**Comment by ihard** [comment]
I will support the question about implementation plans? Or the functionality cannot be implemented for some reason?

**Comment by mophilly** [comment]
> It's not on our todo list at the moment, as we are currently focusing on calls and meetings

What is needed to move this forward? I am wondering about skills, a task list, and how a volunteer could help. Not general, but specific to this topic, please.

**Comment by mophilly** [comment]
I am willing to take a look, but like everyone, my free time is limited. 

If someone could point me to the segments dealing with adding and deleting posts in  talk channels, that would save me a few hours of reading code not related to the edit functions. Perhaps I can help move this forward. 

BTW, the reason I am posting again is I am advocating NC Talk, and get complaint almost every day that a post cannot be edited.

**Comment by edmundlaugasson** [comment]
I would say, that NC team has not decided it as we don't see any official confirmation. Would be reasonable not rely on guesses. As far as I know but not completely sure either, there is lack of developers and possibly other resources.
I truly hope, that this news [https://nextcloud.com/blog/eu-governments-bet-on-nextcloud-to-replace-sharepoint/](https://nextcloud.com/blog/eu-governments-bet-on-nextcloud-to-replace-sharepoint/) brings more money and developers to Nextcloud and it will be developed more diligently and faster. So, thumbs up and keep going!

**Comment by AnAnalogGuy** [comment]
Actually it's not a guess, it's call deduction. Quite simple. Open sind 2019, we have 2023, 4 years, no one picked it up.

**Comment by Gymnae** [comment]
+1
This feature is a must-have. All other platforms have it. 
Maybe another argument why integrating Matrix instead of re-inventing the wheel would've been a good choice.

**Comment by nickvergessen** [comment]
What's app added it now with a 15 minute limit:
https://blog.whatsapp.com/now-you-can-edit-your-whatsapp-messages

**Comment by edmundlaugasson** [comment]
> What's app added it now with a 15 minute limit: https://blog.whatsapp.com/now-you-can-edit-your-whatsapp-messages

Nextcloud should have it without limits. Otherwise people choose something else.

**Comment by q-wertz** [comment]
I like the way Mattermost handles it. The admin can set a limit for how long the message can be edited.

There are scenarios where you do not want that every message is editable for always (and there is not trace what the original message was)

**Comment by edmundlaugasson** [comment]
That's true - this message editing feature should be configurable per chat to fit all needs.

**Comment by nickvergessen** [comment]
Mostlikely we will use the same limitation as for deleting. But more to be decided by the design team.

**Comment by dem36rus** [comment]
+1 need this function

**Comment by NoahDar** [comment]
I too would love to see this happen in Talk.  Just throwing it out there.

**Comment by alauzon** [comment]
I am volunteering to do it if someone leads me into the code.

**Comment by nickvergessen** [comment]
Well, it's kind of a lot of places. Checking the code from the delete could be a good pointer: https://github.com/nextcloud/spreed/pull/4861

But afterwards we need mobile apps to be able to understand the messages, before we can merge the work

**Comment by alauzon** [comment]
Thank you for your answer, Joas.

And when you say "But afterwards we need mobile apps to be able to understand the messages, before we can merge the work" which code is implied in that? The spreed app and/or the Nextcloud Talk app on Android and iPhone?

**Comment by alauzon** [comment]
> Android and iOS apps yeah. They save messages locally, so they need to understand the editing message to update the previous message.

How can we do that? If the messages are saved locally is there a mechanism to tell the phone apps to modify it? Maybe the delete does something similar also?

**Comment by nickvergessen** [comment]
Exactly, there are messages which tell them "hey delete the message with ID x" and for edit we basically need "hey refresh the content of the message with ID x"

**Comment by NoahDar** [comment]
Also, there's been a suggestion that the ability to edit the message should expire after 15 minutes to keep resources down.   Any thoughts?

**Comment by alauzon** [comment]
> Exactly, there are messages which tell them "hey delete the message with ID x" and for edit we basically need "hey refresh the content of the message with ID x"

Do you have the link to the ĥone app(s) ticket(s) for the delete function?

**Comment by nickvergessen** [comment]
Couldn't find an iOS link, Android is here: https://github.com/nextcloud/talk-android/pull/985

**Comment by warnerbryce** [comment]
@alauzon i could give you a tip like 100€ but i can't afford more. tell me if you are interested and where i can tip you

**Comment by christiansass** [comment]
Add me to the waiting list... I'm also trying to see if we can contribute any resources to this issue.

**Comment by jdaviescoates** [comment]
Signal recently got editing and it's _so_ much better now.  I can't wait for Nextcloud Talk to join the "able to edit messages" club too!

**Comment by Folaht** [comment]
Same here.

**Comment by mophilly** [comment]
> Also, there's been a suggestion that the ability to edit the message should expire after 15 minutes to keep resources down. Any thoughts?

It seems to me that is a feature needed when the message volume becomes quite high and impacts the server load. First step is to get a basic edit function in place and well tested.

**Comment by mophilly** [comment]
> Are there people willing to pay me in order to do that? And how can we arrange that? I think that there are tools to help raise money for features, right?

I can contribute a small amount, too. I may be able to assist with some testing.

Please add me to the waiting list

**Comment by iwex** [comment]
Also need this feature

**Comment by nickvergessen** [comment]
Because we don't need further bagging for this feature at the moment, I will lock this ticket.
It's on our radar and might be included in the future, but unless someone can implement it on iOS, Android, WebFrontend and the API level, it will have to wait.

Pointers were given above already:
* API + Frontend https://github.com/nextcloud/spreed/pull/4861
* Android https://github.com/nextcloud/talk-android/pull/985
* iOS https://github.com/nextcloud/talk-ios/pull/486

**Comment by nickvergessen** [comment]
I'm unlocking this due to the comment in #10833 
I also edited the default feature enhancement template into the first post which was missing because it did not exist back then:

---

### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

If you are a developer of either platform you can still anf surely reach out to get involved in implementing this feature.

You could also reach out to help implementing any other feature which then in return frees time to look into things like this feature here which require coordination across all clients.

**Comment by ChildLearningClub** [comment]
Hi, just a thought, but couldn't this be easily piggybacked on the work that has already been done with the delete and copy message functionalities? The "Edit message" option could run the copy message function -> then the "Delete" -> in replace the "Message deleted by you" by the pasted text from copy. Some code would need to be added " if edit pressed: enable "send message" again logic. I realize that the "Copy formatted message" is a system copy to clipboard so that would also need to be adjusted. 

Anyway, just throwing that out as an idea. May be total nonsense, but i guess if it gets some thinking then it is worth the post :).

**Comment by nickvergessen** [comment]
Closed via https://github.com/nextcloud/spreed/issues/11203

**Comment by szaimen** [comment]
🎉🎉🎉🎉🎉
